### PR TITLE
[Consensus 2.0] Calculate committed sub dag digest

### PR DESCRIPTION
**Pull request by akichidis**
## Description 

Currently the default digest is being used on the mysticeti committed subdag making the fork detection in checkpoints weaker as included in the `ConsensusCommitInfo`. This PR is calculating the committed sub dag digest and enabling it behind a feature flag. It has to be noted that the reputation scores are not included in the sub dag digest as during crash recovery we don't restore the actual scores for every committed sub dag, but only restoring the last ones and sent as part of the last committed sub dag. 

## Test plan 

CI

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

**Review by mwtian — COMMENTED** [review]
Great work figuring out why prologue transaction digest was not changing with forked consensus commits!

For the commit digest without reputation score, my suggestion is to use `CommitRef` in place of `CommitIndex` in `CommittedSubDag`, and take the digest from the ref. Then we don't need to maintain the separate digest computation, and there is one less type of digest.

**Comment by akichidis** [comment]
> Great work figuring out why prologue transaction digest was not changing with forked consensus commits!
> 
> For the commit digest without reputation score, my suggestion is to use `CommitRef` in place of `CommitIndex` in `CommittedSubDag`, and take the digest from the ref. Then we don't need to maintain the separate digest computation, and there is one less type of digest. If we ever want to use checkpoint to certify consensus commit, this is the digest to use as well.

Thanks @mwtian . Yes I’ve considered using CommitRef, it was actually my first attempt , but we kind of have a cyclic dependency here. We first construct the sub dag and then the Commit - which calculates its digest . In this case we would have to go back and update the sub dag commit ref which is not ideal. Let me try again and see if there is an elegant way, but those two structs they kind of seem a bit independent and related in the same time.

**Comment by akichidis** [comment]
@mwtian @arun-koshy please take another look after the latest changes

**Review by arun-koshy — APPROVED** [review]
Thanks for taking care of this and the cleanups!

File: consensus/core/src/commit.rs [hunk 1]
```diff
@@ -307,52 +311,52 @@ impl CommittedSubDag {
...
+    })
 }
 
 impl Display for CommittedSubDag {
     fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
         write!(
             f,
             "CommittedSubDag(leader={}, index={}, blocks=[",
-            self.leader, self.commit_index
+            self.leader, self.commit_ref
```
Comment (mwtian) [inline consensus/core/src/commit.rs:345]: Update the message above to `ref={}`?

File: consensus/core/src/commit.rs [hunk 2]
```diff
@@ -307,52 +311,52 @@ impl CommittedSubDag {
...
         )?;
         for (idx, block) in self.blocks.iter().enumerate() {
             if idx > 0 {
                 write!(f, ", ")?;
             }
             write!(f, "{}", block.digest())?;
         }
-        write!(f, "])")
+        write!(f, "], digest={})", self.commit_ref.digest)
```
Comment (mwtian) [inline consensus/core/src/commit.rs:353]: Revert this?

File: consensus/core/src/linearizer.rs [hunk 3]
```diff
@@ -23,15 +24,19 @@ impl Linearizer {
...
+    fn collect_sub_dag_and_commit(
         &mut self,
         leader_block: VerifiedBlock,
-        last_commit_index: CommitIndex,
-        last_commit_timestamp_ms: BlockTimestampMs,
-        last_committed_rounds: Vec<Round>,
-    ) -> CommittedSubDag {
+    ) -> (CommittedSubDag, TrustedCommit) {
+        // Grab latest commit state from dag state
+        let dag_state = self.dag_state.read();
```
Comment (mwtian) [inline consensus/core/src/linearizer.rs:34]: Eventually I hope we can place where the current commit state is read from DagState, and where commits are buffered in DagState, inside the same function. It give readers of the code more confidence that commit state gets updated by the last commit. We can follow up with another PR after the score refactor. In the short term, maybe we can just add an assertion that the last commit index is current index - 1.

**Review by mwtian — APPROVED** [review]

File: crates/sui-protocol-config/src/lib.rs [hunk 4]
```diff
@@ -430,6 +430,12 @@ struct FeatureFlags {
     // Controls leader scoring & schedule change in Mysticeti consensus.
     #[serde(skip_serializing_if = "is_false")]
     mysticeti_leader_scoring_and_schedule: bool,
+
+    // Enables the use of the Mysticeti committed sub dag digest to the `ConsensusCommitInfo` in checkpoints.
+    // When disabled the default digest is used instead. It's important to have this guarded behind
+    // a flag as it will lead to checkpoint forks.
+    #[serde(skip_serializing_if = "is_false")]
+    mysticeti_use_committed_subdag_digest: bool,
```
Comment (mystenmark) [inline crates/sui-protocol-config/src/lib.rs:438]: don't we want to enable this, at least for ChainId::Unknown?

**Review by mystenmark — COMMENTED** [review]
sui-core changes are good, but we should enable it in a new proto version for devnet at least.

having this disabled is actually as big a risk as having it disabled. if we enable it and it causes a fork, that just means that we caught the fork as early as possible

**Comment by mystenmark** [comment]
In fact I would even go further and say I'd really like to see this enabled before we launch on mainnet

File: crates/sui-protocol-config/src/lib.rs (hunk 4 above)
Comment (akichidis) [reply crates/sui-protocol-config/src/lib.rs:438]: oh sure, was just planning to do as separate PR

File: consensus/core/src/linearizer.rs (hunk 3 above)
Comment (akichidis) [reply consensus/core/src/linearizer.rs:34]: > Eventually I hope we can place where the current commit state is read from DagState, and where commits are buffered in DagState, inside the same function. It give readers of the code more confidence that commit state gets updated by the last commit

Hhm I am not sure I understand the above comment. Could you please elaborate a bit more?

Comment (mwtian) [reply consensus/core/src/linearizer.rs:34]: This does not block the PR. But when reading last commit from DagState, it is unintuitive which exact commit is being read unless pushing commits to DagState also happened in the same function, or there is verification that the last commit from DagState indeed has index - 1.
